File: programs in python/min-stack-implementation.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.minstack = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        if len(self.minstack) == 0 or val <= self.minstack[- 1]:
            self.minstack.append(val)

    def pop(self) -> None:
        if len(self.stack) != 0:
            if self.minstack[-1] == self.stack[-1]:
                self.minstack.pop()
            self.stack.pop()

        

    def top(self) -> int:
        if len(self.stack )== 0:
            raise IndexError["stack is empty"]
        return self.stack[-1] 
        

    def getMin(self) -> int:
        if len(self.minstack) == 0 :
            raise IndexError["stack is empty"]
        
        return  self.minstack[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: programs in csharp/min-stack-implementation.cs
public class MinStack {

    private List<int> stack; 
    private List<int> minstack;

    public MinStack() {
        stack = new List<int>(); 
        minstack = new List<int>();
    }
    
    public void Push(int val) { 
        stack.Add(val);
        if(minstack.Count == 0 || val <= minstack[minstack.Count - 1 ]){
            minstack.Add(val);
        }
    }
    
    public void Pop() {
        if(stack.Count != 0){
            if(stack[stack.Count -1 ] == minstack[minstack.Count - 1 ]){
                minstack.RemoveAt(minstack.Count -1);
        }
             stack.RemoveAt(stack.Count - 1);
        }


    }
    
    public int Top() {
        return stack.Count != 0 ? stack[stack.Count - 1] : 0;
    }
    
    public int GetMin() {
        return minstack.Count != 0 ? minstack[minstack.Count - 1] : int.MaxValue;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.Push(val);
 * obj.Pop();
 * int param_3 = obj.Top();
 * int param_4 = obj.GetMin();
 */